### PR TITLE
Add feature to remove unnecessary discards.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/42210

C# 9.0 makes discards non-required in several locations.  So this PR adds an analyzer that fades those locations out and now offers to remove them when unnecessary.

### CODE REVIEW

**Comment by mavasani** [comment]
Adding a documentation note before I forget :-) https://github.com/dotnet/roslyn/blob/master/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                switch (discard.Parent)
+                {
+                    case DeclarationPatternSyntax declarationPattern:
+                        if (declarationPattern.Parent is IsPatternExpressionSyntax isPattern)
+                        {
+                            editor.ReplaceNode(
+                                isPattern,
+                                (current, _) =>
+                                {
+                                    var currentIsPattern = (IsPatternExpressionSyntax)isPattern;
```
Comment (davidwengier) [inline src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:66]: Why is this necessary?

**Review by davidwengier — APPROVED** [review]
It would be ideal, but perhaps not important enough to worry about, if this handled any case that the unused local analyzer is not going to trigger on, which is to say the `_1`, `_2` case, etc.

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:66]: when you have `x is T y` that's an `IsPatternSyntax`.  And an IsPattern is *not* a BinaryExpression.  However, are converting this to `x is T`.  In that case, that's a BinaryExpression.  So we have to specially break this apart and reform into a BinaryExpr. 

Should `x is T` be a BinaryExpr?  Probably not.  But we shipped that way 10 years ago, and so here we are :)

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs [hunk 2]
```diff
@@ -151,6 +151,8 @@ internal static class IDEDiagnosticIds
 
         public const string RemoveRedundantEqualityDiagnosticId = "IDE0100";
 
+        public const string RemoveUnnecessaryDiscardDesignationDiagnosticId = "IDE0110";
```
Comment (mavasani) [inline src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:154]: Why are we spreading out the IDs with holes in the ranges?

Comment (mavasani) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:154]: I notice the same for last 2 rules - I think we should rather keep these sequential as these are opaque IDs. I presume the gaps are introduced to account for similar rules that might come in future and this allows such rules to then slot in these gaps. I don't think we should do this as IDs are opaque numbers, and there are already large number of similar rules which have widely separate IDs.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationTests.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    }
+}",
+                FixedCode = @"
+class C
+{
+    void M(object o)
+    {
+        switch (o)
+        {
+            case int:
```
Comment (mavasani) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationTests.cs:46]: I think there needs to be a follow-up issue/PR to fix another existing IDE code fix (removing unnecessary locals/parameters). For example, https://github.com/dotnet/roslyn/blob/dfb8f6351ac881f133e54b3e3a3f3801e859c01d/src/Analyzers/CSharp/Tests/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs#L2404-L2432

This code fix should be modified to generate `case int:` instead of `case int _:`. @CyrusNajmabadi can you please open a new tracking issue for it?

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+            void Report(DiscardDesignationSyntax discard)
+            {
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    this.Descriptor,
+                    discard.GetLocation(),
+                    ReportDiagnostic.Default,
+                    SpecializedCollections.EmptyEnumerable<Location>(),
+                    ImmutableDictionary<string, string>.Empty));
+            }
```
Comment (mavasani) [inline src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs:70]:
```suggestion
            void Report(DiscardDesignationSyntax discard)
                => context.ReportDiagnostic(Diagnostic.Create(Descriptor, discard.GetLocation());
```
```

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryDiscardDesignation
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp,
+        Name = nameof(RemoveUnnecessaryDiscardDesignationCodeFixProvider)), Shared]
+    internal partial class RemoveUnnecessaryDiscardDesignationCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
```
Comment (mavasani) [inline src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:28]: Not sure if the suppression is required
```suggestion
```

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryDiscardDesignation
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp,
+        Name = nameof(RemoveUnnecessaryDiscardDesignationCodeFixProvider)), Shared]
+    internal partial class RemoveUnnecessaryDiscardDesignationCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```
Comment (mavasani) [inline src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:25]: Consider renaming to `CSharpRemoveUnnecessaryDiscardDesignationCodeFixProvider` for consistency - analyzers/fixers in language-specific layer have the language prefix.

**Review by mavasani — APPROVED** [review]

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 5 above)
Comment (alrz) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:28]: We're using `[Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]` instead.

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:154]: >  and there are already large number of similar rules which have widely separate IDs.

Yes.  And this has been unpleasant :)   For example, not being able to keep the rules organized here in linear order.  Needing them to be spread out randomly in user files, etc.

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs:70]: What is DiagnosticHelper for?

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs:66]: yup!

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs]: done.

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs (hunk 4 above)
Comment (mavasani) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs:70]: > What is DiagnosticHelper for?

For analyzers with code style options, as option itself has a notification/severity that needs to be considered for the effective severity of reported diagnostic. For analyzers without code style option, it should just use the simpler public API to create diagnostics.

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryDiscardDesignation/CSharpRemoveUnnecessaryDiscardDesignationDiagnosticAnalyzer.cs]: done, thanks.

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs]: done.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationTests.cs:46]: Done: https://github.com/dotnet/roslyn/issues/49155

File: src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs (hunk 1 above)
Comment (davidwengier) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs]: I meant why do we need to cast `isPattern` when it looks like it already is the right type, and it doesn't look like the value of it changes, and there is other code in this lambda that uses it instead of `currentIsPattern`.

Not that the explanation you gave wasn't interesting :)

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryDiscardDesignation/RemoveUnnecessaryDiscardDesignationCodeFixProvider.cs]: yeah.  that wa a bug :)  but have fixed in the code that got merged.  sorry i didn't realize it!
